zilliqa/src/transaction.rs

```diff
@@ -0,0 +1,542 @@
+                Address::from_bytes(bytes.into())
+            }
+        };
+        let hash = self.calculate_hash();
+
+        Ok(RecoveredTransaction {
+            tx: self,
+            signer,
+            hash,
+        })
+    }
+
+    /// Calculate the hash of this transaction. If you need to do this more than once, consider caching the result
+    /// using [`Self::recover_signer()`] and the `hash` field from [RecoveredTransaction].
+    pub fn calculate_hash(&self) -> crypto::Hash {
+        match self {
+            SignedTransaction::Legacy { tx, sig } => {
+                let mut rlp = RlpStream::new_list(9);
+                tx.encode_fields(&mut rlp);
+                sig.encode_legacy(&mut rlp, tx.chain_id);
+                crypto::Hash(Keccak256::digest(rlp.out()).into())
+            }
+            SignedTransaction::Eip2930 { tx, sig } => {
+                let mut buffer = BytesMut::with_capacity(1024);
+                buffer.put_u8(1);
+                let mut rlp = RlpStream::new_list_with_buffer(buffer, 11);
+                tx.encode_fields(&mut rlp);
+                sig.encode(&mut rlp);
+                crypto::Hash(Keccak256::digest(rlp.out()).into())
+            }
+            SignedTransaction::Eip1559 { tx, sig } => {
+                let mut buffer = BytesMut::with_capacity(1024);
+                buffer.put_u8(2);
+                let mut rlp = RlpStream::new_list_with_buffer(buffer, 12);
+                tx.encode_fields(&mut rlp);
+                sig.encode(&mut rlp);
+                crypto::Hash(Keccak256::digest(rlp.out()).into())
+            }
+            SignedTransaction::Zilliqa { tx, key, .. } => {
+                let txn_data = encode_zilliqa_transaction(tx, *key);
+                crypto::Hash(Sha256::digest(txn_data).into())
+            }
+        }
+    }
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
+pub struct EthSignature {
+    pub r: [u8; 32],
+    pub s: [u8; 32],
+    // True if the parity of the y value of the signature is odd.
+    pub y_is_odd: bool,
+}
+
+impl EthSignature {
+    fn encode(&self, rlp: &mut RlpStream) {
+        rlp.append(&self.y_is_odd)
+            .append(&strip_leading_zeroes(self.r.as_slice()))
+            .append(&strip_leading_zeroes(self.s.as_slice()));
+    }
+
+    fn encode_legacy(&self, rlp: &mut RlpStream, chain_id: Option<u64>) {
+        // Encode the 'v' value of the signature, based the specification of EIP-155.
+        let v = if let Some(chain_id) = chain_id {
+            (self.y_is_odd as u64) + chain_id * 2 + 35
+        } else {
+            (self.y_is_odd as u64) + 27
+        };
+        rlp.append(&v)
+            .append(&strip_leading_zeroes(self.r.as_slice()))
+            .append(&strip_leading_zeroes(self.s.as_slice()));
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+/// A [SignedTransaction] which has had the signer recovered. The transaction's hash is also calculated and cached.
```

Comment:
Maybe mention the fact that it's intentionally not serializable/deserializable here, for the future